File: DeenTime/backend/DeenTime.Api/Controllers/HijriController.cs
using DeenTime.Core.Entities;
using DeenTime.Infrastructure;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Microsoft.AspNetCore.OutputCaching;
using DeenTime.Core.Services;

namespace DeenTime.Api.Controllers
{
	[ApiController]
	[Authorize]
	[Route("api/v1/[controller]")]
	public sealed class HijriController : ControllerBase
	{
		public record HijriUpsertRequest(Guid OrganizationId, int Year, int Month, int HijriDayOnFirst, bool Locked);
		private readonly AppDbContext _db;
		public HijriController(AppDbContext db) { _db = db; }

		[HttpGet("{orgId:guid}")]
		[OutputCache(PolicyName = "public-read")]
		public async Task<IActionResult> Get(Guid orgId, [FromQuery] string from, [FromQuery] string to)
		{
			if (!DateOnly.TryParse(from + "-01", out var fromDate) || !DateOnly.TryParse(to + "-01", out var toDate))
				return BadRequest("Invalid YYYY-MM range");
			var (fy, fm) = (fromDate.Year, fromDate.Month);
			var (ty, tm) = (toDate.Year, toDate.Month);
			var all = await _db.HijriMonthMaps.AsNoTracking()
				.Where(h => h.OrganizationId == orgId)
				.Where(h => (h.Year > fy || (h.Year == fy && h.Month >= fm)) && (h.Year < ty || (h.Year == ty && h.Month <= tm)))
				.OrderBy(h => h.Year).ThenBy(h => h.Month)
				.ToListAsync();
			return Ok(all);
		}

		[HttpPost]
		[Authorize(Roles = "Admin,Editor")]
		public async Task<IActionResult> Create([FromBody] HijriUpsertRequest req)
		{
			var entity = new HijriMonthMap
			{
				Id = Guid.NewGuid(), OrganizationId = req.OrganizationId,
				Year = req.Year, Month = req.Month, HijriDayOnFirst = req.HijriDayOnFirst, Locked = req.Locked
			};
			_db.HijriMonthMaps.Add(entity);
			await _db.SaveChangesAsync();
			return Created($"/api/v1/hijri/{entity.Id}", entity);
		}

		[HttpPut("{id:guid}")]
		[Authorize(Roles = "Admin,Editor")]
		public async Task<IActionResult> Update(Guid id, [FromBody] HijriUpsertRequest req)
		{
			var existing = await _db.HijriMonthMaps.FirstOrDefaultAsync(h => h.Id == id);
			if (existing is null) return NotFound();
			existing.OrganizationId = req.OrganizationId;
			existing.Year = req.Year;
			existing.Month = req.Month;
			existing.HijriDayOnFirst = req.HijriDayOnFirst;
			existing.Locked = req.Locked;
			existing.UpdatedAtUtc = DateTime.UtcNow;
			await _db.SaveChangesAsync();
			return NoContent();
		}

		[HttpPost("regenerate/{orgId:guid}")]
		[Authorize(Roles = "Admin,Editor")]
		public async Task<IActionResult> Regenerate(Guid orgId, [FromQuery] string from, [FromQuery] string to, [FromServices] IHijriService hijri)
		{
			if (!DateOnly.TryParse(from + "-01", out var f) || !DateOnly.TryParse(to + "-01", out var t)) return BadRequest("Invalid range");
			var existing = await _db.HijriMonthMaps.Where(h => h.OrganizationId == orgId &&
					(h.Year > f.Year || (h.Year == f.Year && h.Month >= f.Month)) &&
					(h.Year < t.Year || (h.Year == t.Year && h.Month <= t.Month)))
					.ToListAsync();
			var locked = existing.Where(x => x.Locked).ToDictionary(x => (x.Year, x.Month));
			_db.HijriMonthMaps.RemoveRange(existing.Where(x => !x.Locked));
			var generated = hijri.Generate(orgId, f, t).Where(m => !locked.ContainsKey((m.Year, m.Month)));
			await _db.HijriMonthMaps.AddRangeAsync(generated);
			await _db.SaveChangesAsync();
			return Ok(new { regenerated = generated.Count(), preservedLocked = locked.Count });
		}
	}
}

File: DeenTime/backend-node/src/modules/hijri/hijri.controller.ts
import { Body, Controller, Get, HttpCode, Param, Post, Put, Query, Res } from '@nestjs/common';
import type { Response } from 'express';
import { randomUUID } from 'node:crypto';
import { z } from 'zod';
import { Authorize, AuthorizeRoles, CurrentUser } from '../../common/auth/authorize.decorator.js';
import { canAccessOrganization } from '../../common/auth/claims.js';
import type { SessionUser } from '../../common/auth/jwt.service.js';
import { badRequestText, forbidden, notFound } from '../../common/errors.js';
import { EMPTY_GUID, formatDateTime, isGuid, parseDateOnly } from '../../common/json.js';
import { validated } from '../../common/zod-validation.pipe.js';
import { generateHijriMonthMaps } from '../../domain/hijri.js';
import { PrismaService } from '../../prisma/prisma.service.js';

interface HijriMonthMapRow {
  id: string;
  organizationId: string;
  year: number;
  month: number;
  hijriDayOnFirst: number;
  hijriMonthOnFirst: number;
  hijriYearOnFirst: number;
  locked: boolean;
  updatedAtUtc: Date;
}

/** HijriMonthMap entity JSON. */
export function toHijriMonthMapJson(row: HijriMonthMapRow): Record<string, unknown> {
  return {
    id: row.id,
    organizationId: row.organizationId,
    year: row.year,
    month: row.month,
    hijriDayOnFirst: row.hijriDayOnFirst,
    hijriMonthOnFirst: row.hijriMonthOnFirst,
    hijriYearOnFirst: row.hijriYearOnFirst,
    locked: row.locked,
    updatedAtUtc: formatDateTime(row.updatedAtUtc),
  };
}

/** HijriUpsertRequest (no validator; model binding defaults). */
const hijriUpsertSchema = z.object({
  organizationId: z.string().nullish().transform((value) => (value && isGuid(value) ? value : EMPTY_GUID)),
  year: z.coerce.number().int().default(0),
  month: z.coerce.number().int().default(0),
  hijriDayOnFirst: z.coerce.number().int().default(0),
  hijriMonthOnFirst: z.coerce.number().int().nullish().transform((value) => value ?? null),
  hijriYearOnFirst: z.coerce.number().int().nullish().transform((value) => value ?? null),
  locked: z.boolean().default(false),
});
type HijriUpsertRequest = z.infer<typeof hijriUpsertSchema>;

function parseMonthRange(from: string | undefined, to: string | undefined, invalidMessage: string): { fromDate: Date; toDate: Date } {
  const fromDate = parseDateOnly(`${from ?? ''}-01`);
  const toDate = parseDateOnly(`${to ?? ''}-01`);
  if (!fromDate || !toDate) throw badRequestText(invalidMessage);
  const months = (toDate.getUTCFullYear() - fromDate.getUTCFullYear()) * 12 + toDate.getUTCMonth() - fromDate.getUTCMonth();
  if (toDate.getTime() < fromDate.getTime() || months > 24) throw badRequestText('Date range must be no more than 24 months.');
  return { fromDate, toDate };
}

function inRange(row: { year: number; month: number }, fromDate: Date, toDate: Date): boolean {
  const key = row.year * 12 + row.month;
  return key >= fromDate.getUTCFullYear() * 12 + fromDate.getUTCMonth() + 1 && key <= toDate.getUTCFullYear() * 12 + toDate.getUTCMonth() + 1;
}

/** Port of HijriController.cs. */
@Controller('api/v1/hijri')
@Authorize()
export class HijriController {
  constructor(private readonly prisma: PrismaService) {}

  @Get(':orgId')
  async get(@CurrentUser() user: SessionUser | null, @Param('orgId') orgId: string, @Query('from') from?: string, @Query('to') to?: string): Promise<Record<string, unknown>[]> {
    if (!isGuid(orgId)) throw notFound();
    if (!canAccessOrganization(user, orgId)) throw forbidden();
    const { fromDate, toDate } = parseMonthRange(from, to, 'Invalid YYYY-MM range');
    const stored = (await this.prisma.hijriMonthMap.findMany({ where: { organizationId: orgId }, orderBy: [{ year: 'asc' }, { month: 'asc' }] })).filter((row) => inRange(row, fromDate, toDate));
    const byMonth = new Map(stored.map((row) => [`${row.year}-${row.month}`, row]));
    const all = [...stored];
    for (const generated of generateHijriMonthMaps(fromDate, toDate)) {
      const existing = byMonth.get(`${generated.year}-${generated.month}`);
      if (!existing) {
        const created = await this.prisma.hijriMonthMap.create({
          data: { id: randomUUID(), organizationId: orgId, ...generated, updatedAtUtc: new Date() },
        });
        all.push(created);
      } else if (!existing.locked && (existing.hijriMonthOnFirst <= 0 || existing.hijriYearOnFirst <= 1)) {
        const updated = await this.prisma.hijriMonthMap.update({
          where: { id: existing.id },
          data: { hijriDayOnFirst: generated.hijriDayOnFirst, hijriMonthOnFirst: generated.hijriMonthOnFirst, hijriYearOnFirst: generated.hijriYearOnFirst, updatedAtUtc: new Date() },
        });
        all[all.indexOf(existing)] = updated;
      }
    }
    return all.sort((a, b) => a.year - b.year || a.month - b.month).map(toHijriMonthMapJson);
  }

  @Post()
  @AuthorizeRoles('Admin', 'Editor')
  async create(@CurrentUser() user: SessionUser | null, @Body(validated(hijriUpsertSchema)) req: HijriUpsertRequest, @Res({ passthrough: true }) response: Response): Promise<Record<string, unknown>> {
    if (!canAccessOrganization(user, req.organizationId)) throw forbidden();
    const created = await this.prisma.hijriMonthMap.create({
      data: {
        id: randomUUID(),
        organizationId: req.organizationId,
        year: req.year,
        month: req.month,
        hijriDayOnFirst: req.hijriDayOnFirst,
        hijriMonthOnFirst: req.hijriMonthOnFirst ?? 1,
        hijriYearOnFirst: req.hijriYearOnFirst ?? 1,
        locked: req.locked,
        updatedAtUtc: new Date(),
      },
    });
    response.status(201).setHeader('Location', `/api/v1/hijri/${created.id}`);
    return toHijriMonthMapJson(created);
  }

  @Put(':id')
  @AuthorizeRoles('Admin', 'Editor')
  @HttpCode(204)
  async update(@CurrentUser() user: SessionUser | null, @Param('id') id: string, @Body(validated(hijriUpsertSchema)) req: HijriUpsertRequest): Promise<void> {
    const existing = isGuid(id) ? await this.prisma.hijriMonthMap.findFirst({ where: { id } }) : null;
    if (!existing) throw notFound();
    if (!canAccessOrganization(user, existing.organizationId)) throw forbidden();
    if (req.organizationId.toLowerCase() !== existing.organizationId.toLowerCase()) throw badRequestText('A Hijri record cannot be moved to another masjid.');
    await this.prisma.hijriMonthMap.update({
      where: { id: existing.id },
      data: {
        year: req.year,
        month: req.month,
        hijriDayOnFirst: req.hijriDayOnFirst,
        hijriMonthOnFirst: req.hijriMonthOnFirst ?? existing.hijriMonthOnFirst,
        hijriYearOnFirst: req.hijriYearOnFirst ?? existing.hijriYearOnFirst,
        locked: req.locked,
        updatedAtUtc: new Date(),
      },
    });
  }

  @Post('regenerate/:orgId')
  @AuthorizeRoles('Admin', 'Editor')
  @HttpCode(200)
  async regenerate(@CurrentUser() user: SessionUser | null, @Param('orgId') orgId: string, @Query('from') from?: string, @Query('to') to?: string): Promise<{ regenerated: number; preservedLocked: number }> {
    if (!isGuid(orgId)) throw notFound();
    if (!canAccessOrganization(user, orgId)) throw forbidden();
    const { fromDate, toDate } = parseMonthRange(from, to, 'Invalid range');
    const existing = (await this.prisma.hijriMonthMap.findMany({ where: { organizationId: orgId } })).filter((row) => inRange(row, fromDate, toDate));
    const locked = new Set(existing.filter((row) => row.locked).map((row) => `${row.year}-${row.month}`));
    const generated = generateHijriMonthMaps(fromDate, toDate).filter((row) => !locked.has(`${row.year}-${row.month}`));
    await this.prisma.$transaction([
      this.prisma.hijriMonthMap.deleteMany({ where: { id: { in: existing.filter((row) => !row.locked).map((row) => row.id) } } }),
      this.prisma.hijriMonthMap.createMany({
        data: generated.map((row) => ({ id: randomUUID(), organizationId: orgId, ...row, updatedAtUtc: new Date() })),
      }),
    ]);
    return { regenerated: generated.length, preservedLocked: locked.size };
  }
}
